File: deep_faceswap_api/src/jobs.rs
//! Job management for long-running operations (video processing)

use serde::Serialize;
use utoipa::ToSchema;

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum JobStatus {
    Queued,
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct JobProgress {
    #[schema(example = "processing_frames")]
    pub stage: String,
    #[schema(example = 150)]
    pub current: usize,
    #[schema(example = 300)]
    pub total: usize,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct JobResult {
    #[schema(example = "/home/user/output.mp4")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_path: Option<String>,
    /// Arbitrary result payload (e.g. video analyze response)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[schema(value_type = Object)]
    pub data: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct JobState {
    #[schema(example = "a1b2c3d4")]
    pub id: String,
    pub status: JobStatus,
    pub progress: JobProgress,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<JobResult>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

impl JobState {
    pub fn new_queued(id: String) -> Self {
        Self {
            id,
            status: JobStatus::Queued,
            progress: JobProgress {
                stage: "queued".to_string(),
                current: 0,
                total: 0,
            },
            result: None,
            error: None,
        }
    }
}

File: frontend/src/lib/stores/jobs.js
import { request } from '../api.js'

export class JobProgress {
	/**
	 * @param {string} stage
	 * @param {number} current
	 * @param {number} total
	 */
	constructor(stage, current, total) {
		/** @type {string} */
		this.stage = stage
		/** @type {number} */
		this.current = current
		/** @type {number} */
		this.total = total
	}
}

export class JobResult {
	/**
	 * @param {string} [output_path]
	 * @param {any} [data]
	 */
	constructor(output_path, data) {
		/** @type {string|undefined} */
		this.output_path = output_path
		/** @type {any} */
		this.data = data
	}
}

export class JobState {
	/**
	 * @param {string} job_id
	 * @param {string} status
	 * @param {JobProgress} progress
	 * @param {JobResult} [result]
	 * @param {string} [error]
	 */
	constructor(job_id, status, progress, result, error) {
		/** @type {string} */
		this.job_id = job_id
		/** @type {string} */
		this.status = status
		/** @type {JobProgress} */
		this.progress = progress
		/** @type {JobResult|undefined} */
		this.result = result
		/** @type {string|undefined} */
		this.error = error
	}
}

/**
 * @param {string} jobId
 * @returns {Promise<JobState>}
 */
export const getJobStatus = async (jobId) => {
	const data = await request('GET', `/jobs/${encodeURIComponent(jobId)}`)
	return new JobState(
		data.id,
		data.status,
		new JobProgress(data.progress.stage, data.progress.current, data.progress.total),
		data.result ? new JobResult(data.result.output_path, data.result.data) : undefined,
		data.error
	)
}
